File: solutions/160.intersection-of-two-linked-lists/intersection-of-two-linked-lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode getIntersectionNode(ListNode headA, ListNode headB) {
        if(headA == null) return null;
        ListNode curA = headA;
        while(curA.next != null){
            curA = curA.next;
        }
        curA.next = headA;
        
        if(headB == null || headB.next == null) {
            curA.next = null;
            return null;
        }
        ListNode slow = headB.next, fast = headB.next.next;
        while(fast != null && fast.next != null){
            if(fast == slow){
                break;
            }
            fast = fast.next.next;
            slow = slow.next;
        }
        if(fast != slow) {
            curA.next = null;
            return null; 
        }
        
        slow = headB;
        while(true){
            if(slow == fast){
                curA.next = null;
                return slow;
            }
            slow = slow.next;
            fast = fast.next;
        }
    }
}

File: solutions/160.intersection-of-two-linked-lists/intersection-of-two-linked-lists.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */

/**
 * @param {ListNode} headA
 * @param {ListNode} headB
 * @return {ListNode}
 */
var getIntersectionNode = function(headA, headB) {
    const lenA = getLength(headA);
    const lenB = getLength(headB);
    const long = lenA > lenB ? lenA : lenB;
    let longHead = lenA > lenB ? headA : headB;
    const short = lenA > lenB ? lenB : lenA;
    let shortHead = lenA > lenB ? headB : headA;
    const diff = long - short;
    for(let i = 0; i < diff; i++){
        longHead = longHead.next;
    }
    while(true){
        if(shortHead === longHead){
            return shortHead;
        }
        shortHead = shortHead.next;
        longHead = longHead.next;
    }
    return null;
    function getLength(head){
        let len = 0;
        while(head !== null){
            len++;
            head = head.next;
        }
        return len;
    }
};